pooch/downloaders.py

```diff
@@ -308,3 +314,113 @@ def callback(data):
             ftp.quit()
             if ispath:
                 output_file.close()
+
+
+class SFTPDownloader:  # pylint: disable=too-few-public-methods
+    """
+    Download manager for fetching files over SFTP.
+
+    When called, downloads the given file URL into the specified local file
+    Uses the :mod:`paramiko` module to manage downloads.
+
+    Use with :meth:`pooch.Pooch.fetch` to customize the download of files
+    (for example, to use authentication or print a progress bar).
+
+    Parameters
+    ----------
+    port : int
+        Port used for the SFTP connection.
+    username : str
+        User name used to login to the server. Only needed if the server
+        requires authentication (i.e., no anonymous SFTP).
+    password : str
+        Password used to login to the server. Only needed if the server
+        requires authentication (i.e., no anonymous SFTP). Use the empty
+        string to indicate no password is required.
+    timeout : int
+        Timeout in seconds for sftp socket operations, use None to mean no
+        timeout.
+    progressbar : bool
+        If True, will print a progress bar of the download to standard
+        error (stderr). Requires `tqdm <https://github.com/tqdm/tqdm>`__ to
+        be installed.
+
+    """
+
+    def __init__(
+        self,
+        port=22,
+        username="anonymous",
+        password="",
+        account="",
+        timeout=None,
+        progressbar=False,
+    ):
+        if paramiko is None:
+            raise ImportError("SFTP is only supported if paramiko is installed")
+
+        self.port = port
+        self.username = username
+        self.password = password
+        self.account = account
+        self.timeout = timeout
+        self.progressbar = progressbar
+        if self.progressbar and tqdm is None:
+            raise ValueError("Missing package 'tqdm' required for progress bars.")
+
+    def __call__(self, url, output_file, pooch):
+        """
+        Download the given URL over SFTP to the given output file.
+
+        Parameters
+        ----------
+        url : str
+            The URL to the file you want to download.
+        output_file : str or file-like object
+            Path (and file name) to which the file will be downloaded.
+        pooch : :class:`~pooch.Pooch`
+            The instance of :class:`~pooch.Pooch` that is calling this
+            method.
+        """
+
+        parsed_url = parse_url(url)
+        isfile = hasattr(output_file, "write")
+
+        connection = paramiko.Transport(sock=(parsed_url["netloc"], self.port))
+
+        if isfile:
+            # paramiko requires a file name as a string, not a file object
+            # file is first closed to avoid any errors and then get name
+            output_file.close()
+            output_file = output_file.name
```

Comment:
@luke-gregor these also don't seem to be covered. Did this error happen in practice? I'm sure if Pooch ever passes an open file to a downloader.

A possible failure for this would be calling `close()` on a file inside a `with` block. Would that cause trouble when exiting the `with`? Some tests for this might be good.

If this situation never happens in normal use of Pooch, it might be best to let paramiko raise the exception and let the user deal with passing a file name instead of open file.

Reply:
Yep, sounds like a neater solution. I've removed this and added a test.